learning_assistant/views.py

```diff
@@ -68,18 +69,24 @@ def post(self, request, course_run_id):
                 data={'detail': 'Learning assistant not enabled for course.'}
             )
 
-        # If user does not have an enrollment record, or is not staff, they should not have access
+        # If user does not have a verified enrollment record, or is not staff, they should not have full access
         user_role = get_user_role(request.user, courserun_key)
         enrollment_object = CourseEnrollment.get_enrollment(request.user, courserun_key)
         enrollment_mode = enrollment_object.mode if enrollment_object else None
         if (
-            (enrollment_mode not in CourseMode.VERIFIED_MODES)
+            # NOTE: Will there ever be a case where the user has a course mod that's
+            # in neither VERIFIED_MODES nor AUDIT_MODES that we need to worry about?
+            enrollment_mode not in CourseMode.VERIFIED_MODES
+            and enrollment_mode in CourseMode.AUDIT_MODES
```

Comment:
I'm not positive, but the logic here seems to match: https://github.com/openedx/frontend-app-learning/blob/6813872dd3b6737a1a6e13764feb9fa7596cbe3d/src/courseware/course/chat/Chat.jsx#L43. So I think that this should cover it?

Reply:
If we want to display the learning assistant to audit learners, doesn't that mean we'll need to change that logic in the frontend as well?

Reply:
The list of [VERIFIED_MODES](https://github.com/openedx/frontend-app-learning/blob/913c8e408662ea5675ed8a27f19afa56b989e44e/src/constants.ts#L40) in the learning MFE is slightly different than the [list in the LMS](https://github.com/openedx/edx-platform/blob/e9ab422101f6e20b07d762b29c75029ffbc2c0b7/common/djangoapps/course_modes/models.py#L208).

I think what is important is that a learner can upgrade from any of the audit modes, right? We'd only want to shower learners than can upgrade the trial. I'm not entirely sure what that exact list is. We may need to figure that out.

Reply:
It seems like the [VERIFIED_MODES](https://github.com/openedx/frontend-app-learning/blob/913c8e408662ea5675ed8a27f19afa56b989e44e/src/constants.ts#L40) in the learning MFE includes all the modes that are not part of the [AUDIT_MODES in the LMS](https://github.com/openedx/edx-platform/blob/e9ab422101f6e20b07d762b29c75029ffbc2c0b7/common/djangoapps/course_modes/models.py#L208), so I'm wondering we should do one of two things:
- Only check is the user's enrollment mode is in AUDIT_MODES,
- or check that the mode is in frontend-app-learning's version of VERIFIED_MODES.

Kind of leaning towards the first option, but let me know your thoughts.

Reply:
I tried to find where we allow learners to upgrade and what modes we allow a learner to be in to upgrade. I can't find a singular place. There is the [is_mode_upgradeable](https://github.com/openedx/edx-platform/blob/e9ab422101f6e20b07d762b29c75029ffbc2c0b7/common/djangoapps/course_modes/models.py#L645) method in the LMS, but it's not that widely used.

There is [serialize_upgrade_info](https://github.com/openedx/edx-platform/blob/82297fa41489d4072d49852d512380c8fc251beb/openedx/core/djangoapps/courseware_api/utils.py#L51), which is used by several of the endpoints the learning MFE uses. [can_show_verified_upgrade](https://github.com/openedx/edx-platform/blob/master/lms/djangoapps/courseware/utils.py#L85) uses [is_mode_upsellable](https://github.com/openedx/edx-platform/blob/master/lms/djangoapps/courseware/utils.py#L55), which refers to `CourseMode.UPSELL_TO_VERIFIED_MODES`, as does `can_show_verified_upgrade` later in teh function. [CourseMode.UPSELL_TO_VERIFIED_MODES](https://github.com/openedx/edx-platform/blob/master/common/djangoapps/course_modes/models.py#L230-L231)
is `[HONOR, AUDIT]`.

This kind of makes sense to me. I don't know if we enroll can enroll learners in ExecEd and Bootcamps via the usual upgrade process, so it's possible learners in the `UNPAID_EXECUTIVE_EDUCATION`, and `UNPAID_BOOTCAMP` modes do not see any upgrade messaging.

If that's the case, we may have three states.
1. learners in verified modes and staff - full access
2. learners in the `audit` and `honor` modes - ability to start trial
3. learners in other modes - no access

I'm harping on this because I wouldn't want to advertise a feature that a learner cannot access via upgrade. I think this is worth a conversation at tomorrow's stand up. What do you think?

P.S. I would also want to confirm that we want all "verified" modes to have access to Xpert. Do we want Masters users to have access to Xpert, for example?

Reply:
I did some digging through the code you linked. Long story short, this seems like what we need (this is a slight modification of what you posted):

1. Any learners in verified modes (per the list in the frontend) or staff will have full access.
2. Any learners with course modes in [CourseMode.UPSELL_TO_VERIFIED_MODES](https://github.com/openedx/edx-platform/blob/master/common/djangoapps/course_modes/models.py#L230-L231) will start or continue their trial (unless it's expired, in which case return 403).
3. Learners in all other course modes (captured in an else statement) will get returned 403.

I've just committed changes reflecting this.